assets/js/modules/analytics/datastore/report.js

```diff
@@ -0,0 +1,151 @@
+				isValidMetrics( metrics ),
+				'Metrics for an Analytics report must be either a string, an array of strings, an object, an array of objects or a mix of strings and objects. If an object is used, it must have "expression" and "alias" properties.',
+			);
+		}
+
+		if ( dimensions ) {
+			invariant(
+				isValidDimensions( dimensions ),
+				'Dimensions for an Analytics report must be either a string, an array of strings, an object, an array of objects or a mix of strings and objects. If an object is used, it must have "name" property.',
+			);
+		}
+
+		if ( orderby ) {
+			invariant(
+				isValidOrders( orderby ),
+				'Orders for an Analytics report must be either an object or an array of objects where each object should have "fieldName" and "sortOrder" properties.',
+			);
+		}
+
+		return { options };
+	},
+} );
+
+const BASE_INITIAL_STATE = {
+	reports: {},
+};
+
+const baseResolvers = {
+	*getReport( options = {} ) {
+		const registry = yield Data.commonActions.getRegistry();
+		const existingReport = registry.select( STORE_NAME ).getReport( options );
+
+		// If there are already alerts loaded in state, consider it fulfilled
+		// and don't make an API request.
+		if ( existingReport ) {
+			return;
+		}
+
+		yield fetchGetReportStore.actions.fetchGetReport( options );
+	},
+};
+
+const baseSelectors = {
+	/**
+	 * Gets a Search Console report for the given options.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object}         state                     Data store's state.
+	 * @param {Object}         options                   Optional. Options for generating the report.
+	 * @param {string}         options.startDate         Required, unless dateRange is provided. Start date to query report data for as YYYY-mm-dd.
+	 * @param {string}         options.endDate           Required, unless dateRange is provided. End date to query report data for as YYYY-mm-dd.
+	 * @param {string}         options.dateRange         Required, alternative to startDate and endDate. A date range string such as 'last-28-days'.
+	 * @param {boolean}        options.compareDateRanges Optional. Only relevant with dateRange. Default false.
+	 * @param {boolean}        options.multiDateRange    Optional. Only relevant with dateRange. Default false.
+	 * @param {Array.<string>} options.metrics           Required. List of metrics to query.
+	 * @param {Array.<string>} options.dimensions        Optional. List of dimensions to group results by. Default an empty array.
+	 * @param {Array.<Object>} options.orderby           Optional. An order definition object, or a list of order definition objects, each one containing 'fieldName' and 'sortOrder'. 'sortOrder' must be either 'ASCENDING' or 'DESCENDING'. Default empty array.
+	 * @param {string}         options.url               Optional. URL to get a report for only this URL. Default an empty string.
+	 * @param {number}         options.limit             Optional. Maximum number of entries to return. Default 1000.
```

Comment:
Let's mark the optional parameters like e.g. `[options.dimensions]`.

Reply:
Updated.